pil-analyzer/src/side_effect_checker.rs

```diff
@@ -0,0 +1,124 @@
+use std::collections::HashMap;
+
+use powdr_ast::{
+    analyzed::{
+        Expression, FunctionValueDefinition, Reference, Symbol, SymbolKind, TypedExpression,
+    },
+    parsed::{FunctionKind, LambdaExpression, StatementInsideBlock},
+};
+
+use lazy_static::lazy_static;
+
+/// Check that query functions are only referenced/defined in a query context
+/// and that constr functions are only referenced/defined in a constr context.
+pub fn check(
+    definitions: &HashMap<String, (Symbol, Option<FunctionValueDefinition>)>,
+    context: FunctionKind,
+    e: &Expression,
+) -> Result<(), String> {
+    SideEffectChecker {
+        definitions,
+        context,
+    }
+    .check(e)
+}
+
+struct SideEffectChecker<'a> {
+    definitions: &'a HashMap<String, (Symbol, Option<FunctionValueDefinition>)>,
+    context: FunctionKind,
+}
+
+impl<'a> SideEffectChecker<'a> {
+    fn check(&mut self, e: &Expression) -> Result<(), String> {
+        match e {
+            Expression::Reference(Reference::Poly(r)) => {
+                let kind = self.function_kind_of_symbol(&r.name);
+                if kind != FunctionKind::Pure && kind != self.context {
+                    return Err(format!(
+                        "Referenced a {kind} function inside a {} context: {}",
+                        self.context, r.name,
+                    ));
+                }
+                Ok(())
+            }
+            Expression::LambdaExpression(LambdaExpression {
+                kind,
+                params: _,
+                body,
+            }) => {
+                if *kind != FunctionKind::Pure && *kind != self.context {
+                    return Err(format!(
+                        "Used a {kind} lambda function inside a {} context: {e}",
+                        self.context
+                    ));
+                }
+                let old_context = self.context;
+                let result = self.check(body);
+                self.context = old_context;
+                result
+            }
+            Expression::BlockExpression(statements, _expr) => {
+                for s in statements {
+                    match s {
+                        StatementInsideBlock::LetStatement(s) => {
+                            if s.value.is_none() && self.context != FunctionKind::Constr {
+                                return Err(format!(
+                                    "Tried to create a witness column in a {} context: {s}",
+                                    self.context
+                                ));
+                            }
+                        }
+                        StatementInsideBlock::Expression(expr) => {
```

Comment:
Why does this imply adding a constraint?

Reply:
It's the same rules as we have on the global level: If it's not a declaration, it must be a constraint:
```
let x; // declaration
let x = f(); // declaration
g(); // must result in a constraint that is added to the set
x = 7; // must result in a constraint that is added to the set
```

Reply:
Note that `StatementInsideBlock` is never the last component of a block, so in 
```
let x = constr || {
  let y = ...;
  let z =  ...;
  f();
  g()
}
```
The last line (`g()`) is the result of the function while `f()` must return a constraint